onnxruntime/core/optimizer/selectors_actions/helpers.h

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+namespace onnxruntime {
+
+//
+// Selection helpers
+//
+
+// Struct to serialize the node indexes in an ORT format model.
+// Use EmptyNodeIndex for nullptr entries in the vectors for missing optional inputs
+struct NodesToOptimizeIndexes {
+  std::vector<NodeIndex> nodes;
+  int num_inputs;
+  int num_extra_variadic_inputs;
+  int num_outputs;
+  int num_extra_variadic_outputs;
+};
+
+// Group of nodes for processing. Accessors are provided for input/target/output nodes.
+// A single variadic input OR output (not both - but no inferencing operator requires that) is currently supported
+class NodesToOptimize {
+ public:
+  enum class NodeType {
+    kInput,   // node providing input to target node
+    kTarget,  // target node
+    kOutput   // node consuming output from target node
+  };
+
+  struct NodeLocation {
+    NodeType type;  // is this a node providing input to the target, or consuming output from it
+    int index;
+  };
+
+  // nodes to assemble. num_inputs and num_outputs default to the size of input_nodes and output_nodes.
+  // specify num_input_defs/num_output_defs if the last input/output is variadic
+  NodesToOptimize(const std::vector<Node*>& input_nodes,
+                  Node& target_node,
+                  const std::vector<Node*>& output_nodes,
+                  int num_input_defs = -1, int num_output_defs = -1);
+
+  // construct from saved NodeIndex values. IsValid() will return false if one or more nodes were missing.
+  // Use EmptyNodeIndex for nullptr entries in the vectors for missing optional inputs
+  NodesToOptimize(Graph& graph, const NodesToOptimizeIndexes& node_indexes);
+
+  static constexpr NodeIndex EmptyNodeIndex = std::numeric_limits<NodeIndex>::max();
+
+  NodesToOptimizeIndexes ToIndexes() const;
+
+  // number of inputs and outputs. these equate to the nodes providing an input/output (defined in the operator schema)
+  // for the target node.
+  // if the target node has a variadic input/output, the nodes providing those will always begin at the last entry
+  // in the input/output nodes (i.e. at num_inputs - 1 or num_outputs - 1).
+  //
+  // e.g if there are 3 inputs (same applies to outputs)
+  // if no variadic input
+  //  num_inputs=3. optional inputs that are missing will have a nullptr entry
+  // else variadic input
+  //   if zero variadic values: num_inputs=3, last input is nullptr
+  //   if one variadic value: num_inputs=3, last input is the single variadic input
+  //   if multiple variadic values: num_inputs=3, total inputs = num_inputs + (NumVariadicInputs() - 1)
+  const int num_inputs;
+  const int num_outputs;
+
+  bool HasVariadicInput() const { return num_extra_variadic_inputs_ > 0; }
+  int NumVariadicInputs() const { return num_extra_variadic_inputs_ + 1; }
```

Comment:
>NumVariadicInputs [](http://example.com/codeflow?start=6&length=17)

if HasVariadicInput() is false, is NumVariadicInputs() 1?

Reply:
good point.